ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB.hs

```diff
@@ -0,0 +1,695 @@
+    "Trying to write to the database with a non-monotonic sequence number: "
+    <> showParen True (shows s1) ""
+    <> " is not <= "
+    <> showParen True (shows s2) ""
+  DbErrInitialisingNonEmpty s ->
+    "The database table that is being initialised is non-empty: " <> s
+  DbErrNoValueHandle vh_id ->
+    "Trying to use non-existing value handle: " <> show vh_id
+  DbErrBadRead ->
+    "Failed to read a value from a database table."
+  DbErrBadRangeRead ->
+    "Failed to read a range of values from a database table."
+  DbErrDirExists path ->
+    "Database directory should not exist already: " <> show path
+  DbErrDirDoesntExist path ->
+    "Database directory should already exist: " <> show path
+  DbErrDirIsNotLMDB path ->
+    "Database directory doesn't contain an LMDB database: "
+    <> show path
+    <> "\nPre-UTxO-HD and In-Memory implementations are incompatible with the LMDB implementation, please delete your ledger database if you want to run with LMDB"
+  DbErrClosed -> "The database has been closed."
+
+{-------------------------------------------------------------------------------
+  Database definition
+-------------------------------------------------------------------------------}
+
+-- | The LMDB database that underlies the backing store.
+data Db m l = Db {
+    -- | The LMDB environment is a pointer to the directory that contains the
+    -- @`Db`@.
+    dbEnv           :: !(LMDB.Environment LMDB.ReadWrite)
+  , dbSettings      :: !(LMDB.Database () DbSettings)
+    -- | The on-disk state of the @`Db`@.
+    --
+    -- The state is kept in an LDMB table with only one key and one value:
+    -- The current sequence number of the @`Db`@.
+  , dbState         :: !(LMDB.Database () DbState)
+    -- | The LMDB tables with the key-value stores.
+  , dbBackingTables :: !(LedgerTables l LMDBMK)
+  , dbFilePath      :: !FilePath
+  , dbTracer        :: !(Trace.Tracer m TraceDb)
+  , dbClosed        :: !(IOLike.TVar m Bool)
+  , dbOpenHandles   :: !(IOLike.TVar m (Map Int (LMDBValueHandle l m)))
+  , dbNextId        :: !(IOLike.TVar m Int)
+  }
+
+newtype LMDBLimits = MkLMDBLimits {unLMDBLimits :: LMDB.Limits}
+
+{-# COMPLETE LMDBLimits #-}
+-- | Configuration to use for LMDB backing store initialisation.
+--
+-- Keep the following in mind:
+-- * @'lmdbMapSize'@ should be a multiple of the OS page size.
+-- * @'lmdbMaxDatabases'@ should be set to at least 2, since the backing store
+--    has 2 internal LMDB databases by default: 1 for the database settings
+--    @'DbSettings'@, and 1 for the database state @'DbState'@.
+pattern LMDBLimits :: Int -> Int -> Int -> LMDBLimits
+pattern LMDBLimits{lmdbMapSize, lmdbMaxDatabases, lmdbMaxReaders} =
+  MkLMDBLimits LMDB.Limits {
+    LMDB.mapSize = lmdbMapSize
+  , LMDB.maxDatabases = lmdbMaxDatabases
+  , LMDB.maxReaders = lmdbMaxReaders
+  }
+
+-- | The database settings.
+--
+-- Todo(jdral): The settings are currently empty, but this may change in the
```

Comment:
@jorisdral What do we want to do with this TODO

Reply:
Let's downgrade it from a TODO to just a regular comment. The comment should describe why it is currently empty, and give some pointers on what could be included in the settings, if we were to ever need that.

Reply:
Done